tdmpc2/tdmpc2.py

```diff
@@ -299,30 +303,48 @@ def update(self, buffer):
             if self.cfg.multitask:
                 task = task.flatten(1, 2)
 
+        # # Compute targets
+        # with torch.no_grad():
+        #     next_z = self.model.encode(obs[1:], task)
+        #     td_targets = self._td_target(next_z, reward, task)
+
+        # # Prepare for update
+        # self.optim.zero_grad(set_to_none=True)
+        # self.model.train()
+
+        # # Latent rollout
+        # zs = torch.empty(
+        #     self.cfg.horizon + 1,
+        #     self.cfg.batch_size,
+        #     self.cfg.latent_dim,
+        #     device=self.device,
+        # )
+        # z = self.model.encode(obs[0], task)
+        # zs[0] = z
+        # consistency_loss = 0
+        # for t in range(self.cfg.horizon):
+        #     z = self.model.next(z, action[t], task)
+        #     consistency_loss += F.mse_loss(z, next_z[t]) * self.cfg.rho**t
+        #     zs[t + 1] = z
         # Compute targets
         with torch.no_grad():
+            zs = self.model.encode(obs, task)
             next_z = self.model.encode(obs[1:], task)
             td_targets = self._td_target(next_z, reward, task)
+        
+        dynamics_input = zs[:-1]
+        
+        dynamics_input = torch.cat([dynamics_input, action], dim=-1) # concatenate action to dynamics_input
+        dynamics_input = dynamics_input.permute(1, 2, 0) # Reshape zs from (L, B, H) -> (B, H, L) for dynamics model input 
+        # Encode history by S4 block
+        hs, s_T = self.model._dynamics(dynamics_input)
 
-        # Prepare for update
-        self.optim.zero_grad(set_to_none=True)
-        self.model.train()
+        mlp_input = hs.permute(2, 0, 1)
+        zs_hat = self.model._mlp(mlp_input)
```

Comment:
I need to read up more on S4 models but this here seems weird. No action item yet, just leaving this note for me to check afterwards.

Reply:
I tried to format input based on S4Block forward function (https://github.com/state-spaces/s4/blob/e757cef57d89e448c413de7325ed5601aceaac13/models/s4/s4.py#L1892-L1898) which is commonly used in "convolution mode" for training.